fvm/evm/handler/handler.go

```diff
@@ -0,0 +1,414 @@
+}
+
+func (h *ContractHandler) getBlockDraftHeight() uint64 {
+	if h.newBlockDraft == nil {
+		h.setupNewBlockDraft()
+	}
+	return h.newBlockDraft.Height
+}
+
+func (h *ContractHandler) commitBlockDraft() {
+	err := h.blockchain.AppendBlock(h.newBlockDraft)
+	handleError(err)
+	h.emitEvent(types.NewBlockExecutedEvent(h.newBlockDraft))
+	// reset it
+	h.newBlockDraft = nil
+}
+
+func (h ContractHandler) checkGasLimit(limit types.GasLimit) {
+	// check gas limit against what has been left on the transaction side
+	if !h.backend.ComputationAvailable(environment.ComputationKindEVMGasUsage, uint(limit)) {
+		handleError(types.ErrInsufficientComputation)
+	}
+}
+
+func (h ContractHandler) meterGasUsage(res *types.Result) {
+	if res != nil {
+		err := h.backend.MeterComputation(environment.ComputationKindEVMGasUsage, uint(res.GasConsumed))
+		handleError(err)
+	}
+}
+
+func (h *ContractHandler) emitEvent(event *types.Event) {
+	// TODO add extra metering for rlp encoding
+	encoded, err := event.Payload.Encode()
+	handleError(err)
+	err = h.backend.EmitFlowEvent(event.Etype, encoded)
+	handleError(err)
+}
+
+func (h *ContractHandler) getBlockContext() types.BlockContext {
+	return types.BlockContext{
+		BlockNumber:            h.getBlockDraftHeight(),
+		DirectCallBaseGasUsage: types.DefaultDirectCallBaseGasUsage,
+	}
+}
+
+type Account struct {
+	isAuthorized bool
+	address      types.Address
+	fch          *ContractHandler
+}
+
+// newAccount creates a new evm account
+func newAccount(fch *ContractHandler, addr types.Address, isAuthorized bool) *Account {
+	return &Account{
+		isAuthorized: isAuthorized,
+		fch:          fch,
+		address:      addr,
+	}
+}
+
+// Address returns the address associated with the bridged account
+func (a *Account) Address() types.Address {
+	return a.address
+}
+
+// Balance returns the balance of this bridged account
+func (a *Account) Balance() types.Balance {
+	ctx := a.fch.getBlockContext()
+
+	blk, err := a.fch.emulator.NewReadOnlyBlockView(ctx)
+	handleError(err)
+
+	bl, err := blk.BalanceOf(a.address)
+	handleError(err)
+
+	balance, err := types.NewBalanceFromAttoFlow(bl)
+	handleError(err)
+	return balance
+}
+
+// Deposit deposits the token from the given vault into the flow evm main vault
+// and update the account balance with the new amount
+func (a *Account) Deposit(v *types.FLOWTokenVault) {
+	cfg := a.fch.getBlockContext()
+	a.fch.checkGasLimit(types.GasLimit(cfg.DirectCallBaseGasUsage))
+
+	blk, err := a.fch.emulator.NewBlockView(cfg)
+	handleError(err)
+
+	call := types.NewDepositCall(
+		a.address,
+		v.Balance().ToAttoFlow(),
+	)
+	res, err := blk.DirectCall(call)
+	a.fch.meterGasUsage(res)
+	handleError(err)
+
+	encoded, callHash := a.captureCall(call)
+	a.fch.emitEvent(
+		types.NewTransactionExecutedEvent(
+			a.fch.getBlockDraftHeight(),
+			encoded,
+			callHash,
+			res,
+		),
+	)
+	newBalance := a.fch.getBlockDraftTotalSupply() + v.Balance().ToAttoFlow().Uint64()
```

Comment:
I guess overflow is not a real issue for us here?

Reply:
uint64 captures overflows.